File: scripts/sync-stripe-customers.py
#!/usr/bin/env python3
# One-time script: sync Stripe customer IDs into allowed_emails
# Run with:
#   python3 scripts/sync-stripe-customers.py

import urllib.request
import urllib.parse
import json
import os

STRIPE_SECRET_KEY        = os.environ.get("STRIPE_SECRET_KEY")
SUPABASE_SERVICE_ROLE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
SUPABASE_URL             = "https://gyskfutmncprqxazgatv.supabase.co"

if not STRIPE_SECRET_KEY:        raise SystemExit("Missing STRIPE_SECRET_KEY")
if not SUPABASE_SERVICE_ROLE_KEY: raise SystemExit("Missing SUPABASE_SERVICE_ROLE_KEY")

def stripe_get(path):
    req = urllib.request.Request(
        f"https://api.stripe.com{path}",
        headers={"Authorization": f"Bearer {STRIPE_SECRET_KEY}"}
    )
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())

def supabase_get(path):
    req = urllib.request.Request(
        f"{SUPABASE_URL}{path}",
        headers={
            "apikey": SUPABASE_SERVICE_ROLE_KEY,
            "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
        }
    )
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())

def supabase_patch(path, data):
    body = json.dumps(data).encode()
    req = urllib.request.Request(
        f"{SUPABASE_URL}{path}",
        data=body,
        method="PATCH",
        headers={
            "apikey": SUPABASE_SERVICE_ROLE_KEY,
            "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal",
        }
    )
    with urllib.request.urlopen(req) as r:
        return r.status

# ── Fetch all Stripe customers (paginated) ───────────────────────────────────
print("Fetching Stripe customers...")
customers = []
starting_after = None
while True:
    params = "limit=100"
    if starting_after:
        params += f"&starting_after={starting_after}"
    data = stripe_get(f"/v1/customers?{params}")
    customers.extend(data["data"])
    if not data["has_more"]:
        break
    starting_after = data["data"][-1]["id"]

print(f"Found {len(customers)} Stripe customers")

# Build email → most-recent customer map
stripe_by_email = {}
for c in customers:
    if not c.get("email"):
        continue
    key = c["email"].lower()
    if key not in stripe_by_email or c["created"] > stripe_by_email[key]["created"]:
        stripe_by_email[key] = c

# ── Fetch allowed_emails ─────────────────────────────────────────────────────
print("Fetching allowed_emails...")
rows = supabase_get("/rest/v1/allowed_emails?select=email,stripe_customer_id")
print(f"Found {len(rows)} allowed emails\n")

updated = skipped = not_found = 0

for row in rows:
    email = row["email"].lower()
    existing = row.get("stripe_customer_id")
    customer = stripe_by_email.get(email)

    if not customer:
        print(f"  No Stripe customer: {email}")
        not_found += 1
        continue

    if existing == customer["id"]:
        skipped += 1
        continue

    print(f"  Updating {email} → {customer['id']}")
    encoded = urllib.parse.quote(email)
    supabase_patch(f"/rest/v1/allowed_emails?email=eq.{encoded}", {"stripe_customer_id": customer["id"]})
    updated += 1

print(f"\nDone. Updated: {updated}, Already set: {skipped}, No Stripe match: {not_found}")

File: scripts/sync-stripe-customers.js
// One-time script: sync Stripe customer IDs into allowed_emails
// Run with:
//   STRIPE_SECRET_KEY=sk_live_... SUPABASE_SERVICE_ROLE_KEY=... node scripts/sync-stripe-customers.js

const STRIPE_SECRET_KEY       = process.env.STRIPE_SECRET_KEY;
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;
const SUPABASE_URL            = "https://gyskfutmncprqxazgatv.supabase.co";

if (!STRIPE_SECRET_KEY)         { console.error("Missing STRIPE_SECRET_KEY");         process.exit(1); }
if (!SUPABASE_SERVICE_ROLE_KEY) { console.error("Missing SUPABASE_SERVICE_ROLE_KEY"); process.exit(1); }

// ── Fetch all Stripe customers (handles pagination) ──────────────────────────
async function fetchAllStripeCustomers() {
  const customers = [];
  let startingAfter = null;

  while (true) {
    const params = new URLSearchParams({ limit: "100" });
    if (startingAfter) params.set("starting_after", startingAfter);

    const res = await fetch(`https://api.stripe.com/v1/customers?${params}`, {
      headers: { Authorization: `Bearer ${STRIPE_SECRET_KEY}` },
    });
    const data = await res.json();
    if (!res.ok) { console.error("Stripe error:", data); process.exit(1); }

    customers.push(...data.data);
    if (!data.has_more) break;
    startingAfter = data.data[data.data.length - 1].id;
  }

  return customers;
}

// ── Fetch all rows from allowed_emails ──────────────────────────────────────
async function fetchAllowedEmails() {
  const res = await fetch(`${SUPABASE_URL}/rest/v1/allowed_emails?select=email,stripe_customer_id`, {
    headers: {
      apikey: SUPABASE_SERVICE_ROLE_KEY,
      Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
    },
  });
  const data = await res.json();
  if (!res.ok) { console.error("Supabase error:", data); process.exit(1); }
  return data;
}

// ── Update stripe_customer_id for a given email ──────────────────────────────
async function updateCustomerId(email, customerId) {
  const res = await fetch(
    `${SUPABASE_URL}/rest/v1/allowed_emails?email=eq.${encodeURIComponent(email)}`,
    {
      method: "PATCH",
      headers: {
        apikey: SUPABASE_SERVICE_ROLE_KEY,
        Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
        "Content-Type": "application/json",
        Prefer: "return=minimal",
      },
      body: JSON.stringify({ stripe_customer_id: customerId }),
    }
  );
  if (!res.ok) {
    const err = await res.text();
    console.error(`  Failed to update ${email}:`, err);
  }
}

// ── Main ─────────────────────────────────────────────────────────────────────
(async () => {
  console.log("Fetching Stripe customers…");
  const customers = await fetchAllStripeCustomers();
  console.log(`Found ${customers.length} Stripe customers`);

  console.log("Fetching allowed_emails…");
  const rows = await fetchAllowedEmails();
  const allowedSet = new Map(rows.map(r => [r.email.toLowerCase(), r.stripe_customer_id]));
  console.log(`Found ${rows.length} allowed emails\n`);

  // Build email → most-recent customer map (some emails may have duplicates)
  const stripeByEmail = new Map();
  for (const c of customers) {
    if (!c.email) continue;
    const key = c.email.toLowerCase();
    // Keep the most recently created customer for this email
    if (!stripeByEmail.has(key) || c.created > stripeByEmail.get(key).created) {
      stripeByEmail.set(key, c);
    }
  }

  let updated = 0, skipped = 0, notFound = 0;

  for (const [email, existing] of allowedSet) {
    const customer = stripeByEmail.get(email);
    if (!customer) {
      console.log(`  No Stripe customer: ${email}`);
      notFound++;
      continue;
    }
    if (existing === customer.id) {
      skipped++;
      continue;
    }
    console.log(`  Updating ${email} → ${customer.id}`);
    await updateCustomerId(email, customer.id);
    updated++;
  }

  console.log(`\nDone. Updated: ${updated}, Already set: ${skipped}, No Stripe match: ${notFound}`);
})();
